File: src/3454.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class SegmentTree:
    def __init__(self, xs: List[int]):
        self.xs = xs
        self.n = len(xs) - 1
        self.count = [0] * (4 * self.n)
        self.covered = [0] * (4 * self.n)

    def update(self, qleft, qright, qval, left, right, pos) -> None:
        if (self.xs[right + 1] <= qleft) or (self.xs[left] >= qright):
            return

        if (qleft <= self.xs[left]) and (self.xs[right + 1] <= qright):
            self.count[pos] += qval
        else:
            middle = (left + right) // 2
            self.update(qleft, qright, qval, left, middle, pos * 2 + 1)
            self.update(qleft, qright, qval, middle + 1, right, pos * 2 + 2)

        if self.count[pos] > 0:
            self.covered[pos] = self.xs[right + 1] - self.xs[left]
        else:
            if left == right:
                self.covered[pos] = 0
            else:
                self.covered[pos] = (self.covered[pos * 2 + 1] + self.covered[pos * 2 + 2])

    def query(self) -> int:
        retVal = self.covered[0]

        return retVal


class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        retVal = 0

        events = []
        xs = set()
        for x, y, l in squares:
            events.append((y, 1, x, x + l))
            events.append((y + l, -1, x, x + l))
            xs.update([x, x + l])
        xs = sorted(xs)
        psegmenttree = SegmentTree(xs)
        events.sort()

        psum = []
        widths = []
        totalarea = 0.0
        previousy = events[0][0]

        # scan: calculate total area and record intermediate states
        for y, delta, xl, xr in events:
            length = psegmenttree.query()
            totalarea += (length * (y - previousy))
            psegmenttree.update(xl, xr, delta, 0, psegmenttree.n - 1, 0)
            # record prefix sums and widths
            psum.append(totalarea)
            widths.append(psegmenttree.query())
            previousy = y

        # calculate the target area (half rounded up)
        target = (totalarea + 1) // 2

        # find the first position greater than or equal to target using binary search
        i = bisect_left(psum, target) - 1

        # get the corresponding area, width, and height
        area = psum[i]
        width = widths[i]
        height = events[i][0]

        retVal = height + (totalarea - area * 2) / (width * 2.0)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for squares in [[[0, 0, 1], [2, 2, 1]], [[0, 0, 2], [1, 1, 1]]]:
            # /* Example
            #  *  Input: squares = [[0,0,1],[2,2,1]]
            #  *  Output: 1.00000
            #  *
            #  *  Input: squares = [[0,0,2],[1,1,1]]
            #  *  Output: 1.00000
            #  */
            logging.info("Input: squares = %s", squares)

            retVal = pSolution.separateSquares(squares)
            logging.info("Output: %.5f", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)

File: src/3454.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <vector>

using namespace std;

class SegmentTree {
   private:
    vector<int> count;
    vector<int> covered;
    vector<int> xs;
    int n;

    void modify(int qleft, int qright, int qval, int left, int right, int pos) {
        if (xs[right + 1] <= qleft || xs[left] >= qright) {
            return;
        }

        if ((qleft <= xs[left]) && (xs[right + 1] <= qright)) {
            count[pos] += qval;
        } else {
            int middle = (left + right) / 2;
            modify(qleft, qright, qval, left, middle, pos * 2 + 1);
            modify(qleft, qright, qval, middle + 1, right, pos * 2 + 2);
        }

        if (count[pos] > 0) {
            covered[pos] = xs[right + 1] - xs[left];
        } else {
            if (left == right) {
                covered[pos] = 0;
            } else {
                covered[pos] = covered[pos * 2 + 1] + covered[pos * 2 + 2];
            }
        }
    }

   public:
    SegmentTree(vector<int>& xs_) : xs(xs_) {
        n = xs.size() - 1;
        count.resize(4 * n, 0);
        covered.resize(4 * n, 0);
    }
    void update(int qleft, int qright, int qval) {
        //
        modify(qleft, qright, qval, 0, n - 1, 0);
    }
    int query() {
        int retVal = covered[0];

        return retVal;
    }
};
class Solution {
   public:
    double separateSquares(vector<vector<int>>& squares) {
        double retVal = 0;

        vector<tuple<int, int, int, int>> events;
        set<int> xsSet;
        for (auto& sq : squares) {
            int x = sq[0];
            int y = sq[1];
            int l = sq[2];
            int xr = x + l;
            events.emplace_back(y, 1, x, xr);
            events.emplace_back(y + l, -1, x, xr);
            xsSet.insert(x);
            xsSet.insert(xr);
        }
        // sort events by y-coordinate
        sort(events.begin(), events.end());
        // discrete coordinates
        vector<int> xs(xsSet.begin(), xsSet.end());
        // initialize the segment tree
        SegmentTree segTree(xs);

        vector<double> psum;
        vector<int> widths;
        double totalarea = 0.0;
        int prev = get<0>(events[0]);

        // scan: calculate total area and record intermediate states
        for (auto& [y, delta, xl, xr] : events) {
            int len = segTree.query();
            totalarea += (1LL * len * (y - prev));
            segTree.update(xl, xr, delta);
            // record prefix sums and widths
            psum.push_back(totalarea);
            widths.push_back(segTree.query());
            prev = y;
        }

        // calculate the target area (half rounded up)
        long long target = (long long)(totalarea + 1) / 2;
        // find the first position greater than or equal to target using binary search
        int i = lower_bound(psum.begin(), psum.end(), target) - psum.begin() - 1;
        // get the corresponding area, width, and height
        double area = psum[i];
        int width = widths[i];
        int height = get<0>(events[i]);

        retVal = height + (totalarea - area * 2) / (width * 2.0);

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<vector<int>> squares;
    };
    vector<subject> testData{{{{0, 0, 1}, {2, 2, 1}}}, {{{0, 0, 2}, {1, 1, 1}}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: squares = [[0,0,1],[2,2,1]]
     *  Output: 1.00000
     *
     *  Input: squares = [[0,0,2],[1,1,1]]
     *  Output: 1.00000
     */

    Solution cSolution;
    double answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: squares = [";
        for (long unsigned int j = 0; j < testData[i].squares.size(); ++j) {
            cout << (j == 0 ? "" : ",") << "[";
            for (long unsigned int k = 0; k < testData[i].squares[j].size(); ++k) {
                cout << (k == 0 ? "" : ",") << testData[i].squares[j][k];
            }
            cout << "]";
        }
        cout << "]\n";

        answer = cSolution.separateSquares(testData[i].squares);
        cout.setf(ios::fixed);
        cout.precision(5);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}
